File: trust/src/bin/test_minimax.rs
use anyhow::Result;
use core::f32;
use im::{vector, Vector};
use std::time::Instant;
use trust::tron;

use trust::{
    minimax::{minimax_alpha_beta_eval_all, MinimaxContext, MinimaxResult},
    models::TractModel,
    tron::{Direction, GameState, GameStatus, Player, StatusInfo},
};

fn main() {
    let start = Instant::now();
    let _ = run_sims().expect("bruh");

    let duration = start.elapsed();
    println!("Duration: {:?}", duration);
}

fn run_sims() -> Result<()> {
    let p1 = Player {
        row: 2,
        col: 2,
        can_move: true,
    };

    let p2 = Player {
        row: 7,
        col: 7,
        can_move: true,
    };



    let tract_model: TractModel =
        TractModel::new("tron_model_v2.onnx").expect("failed to init tract model.");

    println!("starting sims");
    let n_games = 100;
    let start = Instant::now();

    let mut p1_wins = 0;
    let mut p2_wins = 0;
    let mut ties = 0;

    for i in 0..n_games {

        let start_players: Vector<Player> = vector![p1, p2];
        let mut curr_game = tron::new_game(start_players, 10, 10);

        let mut curr_status: StatusInfo = tron::get_status(&curr_game);

        while curr_status.status == GameStatus::InProgress {
            // bot chooses player directions

            let p1_mm_result: MinimaxResult = minimax_alpha_beta_eval_all(
                &curr_game,
                4,
                true,
                f32::NEG_INFINITY,
                f32::INFINITY,
                None,
                &MinimaxContext {
                    model: &tract_model,
                    maximizing_player: 0,
                    minimizing_player: 1,
                },
            );

            let p2_mm_result: MinimaxResult = minimax_alpha_beta_eval_all(
                &curr_game,
                4,
                true,
                f32::NEG_INFINITY,
                f32::INFINITY,
                None,
                &MinimaxContext {
                    model: &tract_model,
                    maximizing_player: 1,
                    minimizing_player: 0,
                },
            );

            let p1_direction: Direction = p1_mm_result.principal_variation.unwrap_or(Direction::Up);
            let p2_direction: Direction = p2_mm_result.principal_variation.unwrap_or(Direction::Up);

            let directions: [Direction; 2] = [p1_direction, p2_direction];
            curr_game = tron::next(&curr_game, &directions);
            curr_status = tron::get_status(&curr_game);
        }

        println!("{i}");
        if curr_status.status == GameStatus::Tie {
            ties += 1;
            println!("Tie");
        } else {
            let winner_index = curr_status
                .winner_index
                .expect("Winner index should be set");
            if winner_index == 0 {
                p1_wins += 1;
                println!("P1 win!");
            } else if winner_index == 1 {
                p2_wins += 1;
                println!("P2 win!");
            }
        }
    }

    // println!("Game completed!: {:?}", game.status);

    let duration = start.elapsed();
    println!("Time to run {} games: {:?}", n_games, duration);

    println!("P1 wins: {}, p2 wins: {}, ties: {}", p1_wins, p2_wins, ties);

    Ok(())
}

File: tron-python/tests/test_minimax.py
import sys

import numpy as np

from tron.game import GameState, Player, GameStatus, StatusInfo, next, get_status, Direction
from tron.ai.tron_model import RandomTronModel
from tron.ai.minimax import minimax_alpha_beta_eval_all, basic_minimax, MinimaxContext
from tron.ai.algos import choose_direction_random


def test_minimax():



    tron_model = RandomTronModel()
    context = MinimaxContext(tron_model, maximizing_player=0, minimizing_player=1)


    for i in range(100):

        for depth in range(2):


            game = GameState.new_game(num_players=2, num_rows=10, num_cols=10, random_starts=True)
            game_status = get_status(game).status

            while game_status == GameStatus.IN_PROGRESS:

                basic_mm_result = basic_minimax(game, depth=depth, is_maximizing_player=True, context=context)
                ab_mm_result = minimax_alpha_beta_eval_all(game, depth=depth, is_maximizing_player=True, context=context)


                # print(f"{basic_mm_result.principal_variation=}, {ab_mm_result.principal_variation=}")

                assert basic_mm_result.evaluation == ab_mm_result.evaluation
                if basic_mm_result.principal_variation != ab_mm_result.principal_variation:
                    assert (basic_mm_result.evaluation / 1000.0).is_integer()
                    print(f"{basic_mm_result.evaluation=}, {ab_mm_result.evaluation=}")



            
                p1_dir = Direction.UP if basic_mm_result.principal_variation is None else basic_mm_result.principal_variation
                p2_dir = choose_direction_random(game, player_index=1)

                game = next(
                    game, directions=(p1_dir, p2_dir)
                )

                game_status = get_status(game).status



